virtual_rainforest/models/plants/functions.py

```diff
@@ -0,0 +1,170 @@
+"""The :mod:`~virtual_rainforest.models.plants.functions` submodule provides the core
+functions used to estimate the canopy model, productivity and growth.
+
+NOTE - much of this will be outsourced to pyrealm.
+
+"""  # noqa: D205, D415
+
+from __future__ import annotations
+
+import numpy as np
+from numpy.typing import NDArray
+from xarray import DataArray
+
+from virtual_rainforest.core.data import Data
+from virtual_rainforest.core.exceptions import ConfigurationError, InitialisationError
+from virtual_rainforest.core.logger import LOGGER
+from virtual_rainforest.core.utils import set_layer_roles
+from virtual_rainforest.models.plants.community import PlantCohort, PlantCommunities
+
+
+def generate_canopy_model(community: list[PlantCohort]) -> tuple[NDArray, NDArray]:
+    """Generate the canopy structure for a plant community.
+
+    This function takes a list of plant cohorts present in a community and uses the T
+    Model to estimate the heights and crown areas of the individuals. It then uses the
+    perfect plasticity approximation to calculate the closure heights of the canopy
+    layers and the leaf area indices of each layer.
+
+    Args:
+        community: A list of plant cohorts.
+
+    Returns:
+        A tuple of one dimensional numpy arrays giving the canopy layer heights and leaf
+        area indices.
+    """
+
+    # TODO - actually calculate these
+    layer_hght = np.array([30.0, 20.0, 10.0])
+    layer_lai = np.array([1.0, 1.0, 1.0])
+
+    return layer_hght, layer_lai
+
+
+def build_canopy_arrays(
+    communities: PlantCommunities, n_canopy_layers: int
+) -> tuple[DataArray, DataArray]:
+    """Converts the PlantCommunities data into canopy layer data arrays.
+
+    This function takes a list of plant cohorts present in a community and uses the T
+    Model to estimate the heights and crown areas of the individuals. It then uses the
+    perfect plasticity approximation to calculate the closure heights of the canopy
+    layers and the leaf area indices of each layer.
+
+    Args:
+        communities: The PlantCommunities object to convert
+        n_canopy_layers: The maximum number of permitted canopy layers.
+
+    Returns:
+        A tuple of two dimensional numpy arrays giving the canopy layer heights and leaf
+        area indices by cell id.
+    """
+
+    # TODO - this could be a method of PlantCommunities but creates circular import of
+    #        PlantCohorts
+
+    # Initialise list of arrays
+    layer_heights: list = []
+    layer_lai: list = []
+    cell_has_too_many_layers: list = []
+
+    # Loop over the communities in each cell
+    for cell_id, community in communities.items():
+        # Calculate the canopy model for the cell and pad as needed
+        this_lyr_hght, this_lyr_lai = generate_canopy_model(community)
```

Comment:
Couldn't this just be `layer_height`, `layer_lai`? as I don't think these are used elsewhere in the function